# Scraping Yahoo Finance

In this assignment you are required to look at historical data for 30 companies from [Dow Jones Index](https://en.wikipedia.org/wiki/Dow_Jones_Industrial_Average). Tickers for companies from the index can be found in *dow_jones_tickers.txt*. For each company you should get historical daily stock prices for 2019 from https://finance.yahoo.com/, and then use the data to answer the questions you will find below. 

In [106]:
dow_jones_companies = []

with open('dow_jones_tickers.txt') as f:
    for ticker in f:
        dow_jones_companies.append(ticker.strip())
N = len(dow_jones_companies)  


In [168]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
data = {}
for ticker in dow_jones_companies:
    url = 'https://query1.finance.yahoo.com/v7/finance/download/' + ticker + '?period1=1546300800&period2=1577750400&interval=1d&events=history&includeAdjustedClose=true'
    tab = pd.read_csv(url)
    tab['Date']= pd.to_datetime(tab['Date'])
    data[ticker] = {'table':tab}
    print(ticker)
    display(data[ticker]['table'].head(-1))

AXP


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   93.910004   96.269997   93.769997   95.680000   92.643311   
1   2019-01-03   94.790001   95.180000   93.230003   93.430000   90.834969   
2   2019-01-04   95.370003   98.349998   94.900002   97.639999   94.928040   
3   2019-01-07   97.769997   98.919998   97.019997   98.169998   95.443306   
4   2019-01-08   98.690002   99.519997   97.519997   98.650002   95.909988   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  125.860001  126.279999  124.830002  125.769997  123.564117   
246 2019-12-23  125.739998  125.739998  123.699997  124.489998  122.306564   
247 2019-12-24  124.949997  125.330002  124.379997  124.739998  122.552185   
248 2019-12-26  124.980003  125.440002  124.529999  125.410004  123.210434   
249 2019-12-27  125.839996  125.970001  125.110001  125.190002  122.994301   

      Volume  
0    4175400  
1    4776600  
2    4637200  
3    3289000  
4    3227200  
..       ...  
245  5828700  
246  3031400  
247   953500  
248  1486600  
249  1788600  

[250 rows x 7 columns]

AAPL


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  38.722500  39.712502  38.557499  39.480000  38.249401   
1   2019-01-03  35.994999  36.430000  35.500000  35.547501  34.439476   
2   2019-01-04  36.132500  37.137501  35.950001  37.064999  35.909672   
3   2019-01-07  37.174999  37.207500  36.474998  36.982498  35.829746   
4   2019-01-08  37.389999  37.955002  37.130001  37.687500  36.512772   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  70.557503  70.662498  69.639999  69.860001  68.699387   
246 2019-12-23  70.132500  71.062500  70.092499  71.000000  69.820442   
247 2019-12-24  71.172501  71.222504  70.730003  71.067497  69.886818   
248 2019-12-26  71.205002  72.495003  71.175003  72.477501  71.273392   
249 2019-12-27  72.779999  73.492500  72.029999  72.449997  71.246353   

        Volume  
0    148158800  
1    365248800  
2    234428400  
3    219111200  
4    164101200  
..         ...  
245  275978000  
246   98572000  
247   48478800  
248   93121200  
249  146266000  

[250 rows x 7 columns]

BA


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  316.190002  323.950012  313.709991  323.809998  314.645142   
1   2019-01-03  319.489990  319.739990  309.399994  310.899994  302.100555   
2   2019-01-04  316.690002  328.440002  316.079987  327.079987  317.822601   
3   2019-01-07  330.519989  330.690002  323.350006  328.109985  318.823395   
4   2019-01-08  334.109985  341.250000  332.649994  340.529999  330.891937   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  334.619995  335.500000  328.000000  328.000000  326.060028   
246 2019-12-23  336.750000  340.589996  334.200012  337.549988  335.553558   
247 2019-12-24  339.510010  339.750000  332.619995  333.000000  331.030457   
248 2019-12-26  332.700012  333.359985  329.299988  329.920013  327.968689   
249 2019-12-27  330.200012  333.649994  329.170013  330.140015  328.187408   

      Volume  
0    3292200  
1    5705600  
2    4448800  
3    4030300  
4    4736500  
..       ...  
245  9489400  
246  9534800  
247  4120100  
248  4593400  
249  4124000  

[250 rows x 7 columns]

CAT


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  124.029999  127.879997  123.000000  126.379997  119.302582   
1   2019-01-03  125.309998  125.500000  120.440002  121.510002  114.705315   
2   2019-01-04  124.169998  128.589996  123.699997  128.149994  120.973465   
3   2019-01-07  128.589996  129.600006  126.699997  128.229996  121.048973   
4   2019-01-08  129.899994  131.500000  128.229996  129.770004  122.502747   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  147.380005  147.770004  145.240005  147.570007  143.254532   
246 2019-12-23  147.679993  148.529999  146.779999  148.500000  144.157303   
247 2019-12-24  148.699997  148.800003  147.289993  147.479996  143.167160   
248 2019-12-26  147.899994  148.229996  146.990005  148.220001  143.885513   
249 2019-12-27  148.639999  148.929993  147.990005  148.279999  143.943756   

      Volume  
0    4783200  
1    6165800  
2    5456600  
3    4653400  
4    4202600  
..       ...  
245  6805400  
246  2346500  
247   966200  
248  1576900  
249  1544500  

[250 rows x 7 columns]

CSCO


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  42.279999  43.200001  42.209999  42.950001  40.382099   
1   2019-01-03  42.299999  42.410000  40.959999  41.070000  38.913486   
2   2019-01-04  41.849998  43.000000  41.529999  42.919998  40.666344   
3   2019-01-07  42.860001  43.650002  42.759998  43.209999  40.941113   
4   2019-01-08  43.570000  43.939999  43.099998  43.560001  41.272739   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  48.250000  48.369999  47.419998  47.450001  45.873497   
246 2019-12-23  47.549999  48.299999  47.419998  48.099998  46.501900   
247 2019-12-24  48.160000  48.250000  47.650002  47.779999  46.192528   
248 2019-12-26  47.779999  48.000000  47.650002  47.849998  46.260204   
249 2019-12-27  47.990002  48.020000  47.680000  47.770000  46.182861   

       Volume  
0    23833500  
1    29606700  
2    27515000  
3    19488100  
4    22071400  
..        ...  
245  52575300  
246  20253500  
247   7044700  
248  11091700  
249  11056800  

[250 rows x 7 columns]

CVX


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  107.339996  111.680000  107.099998  110.690002  100.664474   
1   2019-01-03  111.059998  111.290001  108.070000  108.570000   98.736496   
2   2019-01-04  110.279999  110.989998  109.470001  110.820000  100.782707   
3   2019-01-07  111.059998  112.930000  109.919998  112.260002  102.092270   
4   2019-01-08  113.400002  113.790001  111.669998  111.769997  101.646660   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  119.400002  119.940002  118.089996  119.680000  113.262314   
246 2019-12-23  119.309998  120.400002  119.150002  120.330002  113.877464   
247 2019-12-24  120.430000  120.779999  120.089996  120.339996  113.886909   
248 2019-12-26  120.669998  121.419998  120.339996  120.599998  114.132973   
249 2019-12-27  120.889999  121.290001  120.239998  120.300003  113.849068   

       Volume  
0     6384800  
1     6496200  
2     7952500  
3     5712700  
4     5674900  
..        ...  
245  12171100  
246   5875000  
247   1640100  
248   3648600  
249   4138300  

[250 rows x 7 columns]

XOM


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  67.349998  69.870003  67.260002  69.690002  61.586319   
1   2019-01-03  70.000000  70.250000  68.360001  68.620003  60.640751   
2   2019-01-04  69.639999  71.269997  69.360001  71.150002  62.876545   
3   2019-01-07  71.250000  72.099998  70.739998  71.519997  63.203518   
4   2019-01-08  72.540001  72.589996  71.940002  72.040001  63.663055   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  69.860001  70.230003  69.389999  69.940002  64.767502   
246 2019-12-23  69.860001  70.300003  69.529999  70.290001  65.091606   
247 2019-12-24  70.349998  70.500000  69.910004  70.019997  64.841568   
248 2019-12-26  70.190002  70.500000  70.010002  70.129997  64.943436   
249 2019-12-27  70.199997  70.309998  69.879997  69.889999  64.721184   

       Volume  
0    16727200  
1    13866100  
2    16043600  
3    10844200  
4    11439000  
..        ...  
245  35092000  
246  14172500  
247   3979400  
248   8840200  
249  10516100  

[250 rows x 7 columns]

GS


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  164.330002  172.250000  163.350006  172.029999  165.488129   
1   2019-01-03  170.660004  171.770004  168.289993  169.509995  163.063950   
2   2019-01-04  172.990005  176.000000  171.100006  175.050003  168.393295   
3   2019-01-07  175.229996  177.830002  172.270004  176.020004  169.326401   
4   2019-01-08  177.779999  178.100006  172.899994  175.369995  168.701111   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  230.960007  231.289993  228.619995  228.929993  224.764435   
246 2019-12-23  229.660004  230.479996  228.410004  229.089996  224.921524   
247 2019-12-24  228.979996  229.949997  228.369995  229.910004  225.726639   
248 2019-12-26  230.259995  232.199997  229.990005  231.210007  227.002960   
249 2019-12-27  231.809998  232.210007  230.470001  230.660004  226.462967   

      Volume  
0    3999400  
1    4060200  
2    3788300  
3    3152100  
4    2692700  
..       ...  
245  3715500  
246  1670400  
247   467700  
248  1704000  
249  1343900  

[250 rows x 7 columns]

HD


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  169.710007  173.179993  169.050003  172.410004  164.864197   
1   2019-01-03  171.839996  171.839996  168.210007  168.610001  161.230515   
2   2019-01-04  170.179993  174.740005  169.520004  173.619995  166.021240   
3   2019-01-07  173.830002  178.179993  173.830002  177.039993  169.291534   
4   2019-01-08  178.570007  179.589996  175.610001  177.889999  170.104340   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  220.589996  221.570007  219.800003  221.190002  217.270020   
246 2019-12-23  222.000000  222.000000  219.139999  219.210007  215.325104   
247 2019-12-24  219.500000  221.039993  219.220001  220.669998  216.759232   
248 2019-12-26  220.740005  221.149994  219.589996  220.820007  216.906586   
249 2019-12-27  221.199997  221.789993  219.779999  219.970001  216.071625   

      Volume  
0    4242900  
1    4810500  
2    4762600  
3    4928800  
4    4722400  
..       ...  
245  7778200  
246  3943700  
247  2260700  
248  2649200  
249  3082800  

[250 rows x 7 columns]

IBM


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  112.010002  115.980003  111.690002  115.209999  104.458885   
1   2019-01-03  114.529999  114.879997  112.690002  112.910004  102.373520   
2   2019-01-04  114.910004  117.489998  114.440002  117.320000  106.371994   
3   2019-01-07  117.500000  118.830002  116.669998  118.150002  107.124527   
4   2019-01-08  119.660004  120.570000  118.980003  119.830002  108.647758   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  135.740005  136.419998  134.949997  135.589996  128.833481   
246 2019-12-23  135.779999  136.149994  135.000000  135.550003  128.795471   
247 2019-12-24  135.610001  135.619995  134.610001  134.979996  128.253876   
248 2019-12-26  134.979996  135.309998  134.649994  134.910004  128.187378   
249 2019-12-27  135.000000  135.750000  134.869995  135.270004  128.529419   

      Volume  
0    4239900  
1    4346700  
2    4477800  
3    3751200  
4    4763600  
..       ...  
245  7111800  
246  2803200  
247  1202100  
248  2129700  
249  2752200  

[250 rows x 7 columns]

INTC


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  45.959999  47.470001  45.770000  47.080002  44.819595   
1   2019-01-03  46.150002  46.279999  44.389999  44.490002  42.353947   
2   2019-01-04  45.840000  47.570000  45.540001  47.220001  44.952866   
3   2019-01-07  47.099998  48.000000  46.750000  47.439999  45.162308   
4   2019-01-08  47.799999  48.029999  46.779999  47.740002  45.447910   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  58.320000  59.049999  57.889999  58.950001  57.519756   
246 2019-12-23  59.060001  59.779999  59.040001  59.230000  57.792961   
247 2019-12-24  59.250000  59.480000  59.160000  59.410000  57.968597   
248 2019-12-26  59.410000  59.849998  59.349998  59.820000  58.368649   
249 2019-12-27  59.910000  60.480000  59.900002  60.080002  58.622341   

       Volume  
0    18774600  
1    32267300  
2    35447300  
3    22736800  
4    22749200  
..        ...  
245  49953000  
246  23593000  
247   6313200  
248  11480300  
249  14085800  

[250 rows x 7 columns]

JNJ


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  128.130005  128.380005  126.279999  127.750000  121.799690   
1   2019-01-03  128.139999  128.270004  125.360001  125.720001  119.864243   
2   2019-01-04  127.120003  128.649994  126.730003  127.830002  121.875969   
3   2019-01-07  127.629997  128.350006  126.800003  127.010002  121.094154   
4   2019-01-08  128.179993  130.500000  127.730003  129.960007  123.906761   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  146.880005  147.839996  144.800003  146.059998  143.182373   
246 2019-12-23  145.740005  146.759995  145.339996  146.440002  143.554901   
247 2019-12-24  146.000000  146.009995  145.259995  145.929993  143.054947   
248 2019-12-26  145.970001  146.179993  145.160004  145.830002  142.956924   
249 2019-12-27  146.199997  146.550003  145.539993  145.750000  142.878494   

       Volume  
0     7631700  
1     8654500  
2     8831700  
3     8404700  
4     9351600  
..        ...  
245  14532800  
246   6558700  
247   4238900  
248   3019300  
249   3432700  

[250 rows x 7 columns]

KO


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  46.939999  47.220001  46.560001  46.930000  44.328362   
1   2019-01-03  46.820000  47.369999  46.529999  46.639999  44.054436   
2   2019-01-04  46.750000  47.570000  46.639999  47.570000  44.932877   
3   2019-01-07  47.570000  47.750000  46.900002  46.950001  44.347252   
4   2019-01-08  47.250000  47.570000  47.040001  47.480000  44.847870   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  54.939999  55.200001  54.470001  54.970001  53.567886   
246 2019-12-23  55.000000  55.240002  54.840000  54.910000  53.509411   
247 2019-12-24  54.799999  55.000000  54.639999  54.709999  53.314514   
248 2019-12-26  54.919998  55.020000  54.790001  55.020000  53.616611   
249 2019-12-27  55.009998  55.450001  55.000000  55.349998  53.938194   

       Volume  
0    11603700  
1    14714400  
2    13013700  
3    13135500  
4    15420700  
..        ...  
245  23832800  
246   9300800  
247   3359300  
248   6228500  
249   6895500  

[250 rows x 7 columns]

JPM


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   95.949997   99.779999   95.940002   99.309998   92.928543   
1   2019-01-03   98.639999   98.889999   96.690002   97.110001   91.607857   
2   2019-01-04   99.110001  100.930000   98.279999  100.690002   94.985039   
3   2019-01-07  100.430000  101.470001   99.709999  100.760002   95.051064   
4   2019-01-08  101.629997  101.820000   99.550003  100.570000   94.871834   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  138.020004  138.880005  136.899994  137.240005  132.430801   
246 2019-12-23  137.600006  137.910004  136.880005  137.199997  132.392197   
247 2019-12-24  137.330002  137.630005  137.059998  137.580002  132.758896   
248 2019-12-26  137.889999  139.100006  137.779999  139.039993  134.167709   
249 2019-12-27  139.300003  139.770004  138.669998  139.139999  134.264221   

       Volume  
0    15670900  
1    16286400  
2    16935200  
3    15430700  
4    13578800  
..        ...  
245  29992600  
246   9738400  
247   3324300  
248   6770900  
249   7868200  

[250 rows x 7 columns]

MCD


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  175.410004  176.300003  174.169998  176.059998  168.723358   
1   2019-01-03  175.449997  176.449997  174.410004  174.899994  167.611679   
2   2019-01-04  176.029999  179.199997  175.690002  178.279999  170.850830   
3   2019-01-07  178.139999  181.880005  178.139999  180.220001  172.710022   
4   2019-01-08  181.960007  182.289993  179.910004  180.600006  173.074173   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  198.550003  199.000000  196.649994  197.139999  193.484451   
246 2019-12-23  197.539993  197.539993  196.029999  196.199997  192.561874   
247 2019-12-24  195.929993  197.139999  195.669998  196.669998  193.023178   
248 2019-12-26  196.610001  197.160004  196.160004  197.059998  193.405945   
249 2019-12-27  197.330002  198.470001  197.179993  198.169998  194.495361   

      Volume  
0    2571500  
1    3728200  
2    3194300  
3    3283000  
4    2941400  
..       ...  
245  6415500  
246  2388900  
247  1152500  
248  1809600  
249  2263900  

[250 rows x 7 columns]

MMM


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  187.820007  190.990005  186.699997  190.949997  178.094437   
1   2019-01-03  188.279999  188.279999  182.889999  183.759995  171.388519   
2   2019-01-04  186.750000  191.979996  186.029999  191.320007  178.439545   
3   2019-01-07  191.360001  192.300003  188.660004  190.880005  178.029175   
4   2019-01-08  193.000000  194.110001  189.580002  191.679993  178.775299   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  174.789993  175.470001  172.199997  175.369995  169.074448   
246 2019-12-23  176.710007  179.179993  176.559998  178.470001  172.063171   
247 2019-12-24  178.270004  178.669998  176.039993  176.679993  170.337433   
248 2019-12-26  176.449997  176.660004  175.500000  176.589996  170.250656   
249 2019-12-27  176.649994  177.539993  175.619995  177.259995  170.896606   

      Volume  
0    2475200  
1    3358200  
2    2995100  
3    2162200  
4    2479800  
..       ...  
245  7513000  
246  4876300  
247  1259500  
248  1435200  
249  1472900  

[250 rows x 7 columns]

MRK


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  75.290001  75.989998  74.599998  75.589996  71.884117   
1   2019-01-03  75.559998  75.970001  73.849998  74.040001  70.410133   
2   2019-01-04  74.739998  76.500000  74.639999  76.269997  72.530785   
3   2019-01-07  76.150002  76.470001  75.410004  75.430000  71.731987   
4   2019-01-08  75.870003  76.209999  75.290001  75.989998  72.264519   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  91.459999  92.639999  90.839996  91.580002  89.462669   
246 2019-12-23  91.830002  92.019997  91.440002  91.739998  89.618973   
247 2019-12-24  91.639999  91.730003  91.070000  91.440002  89.325905   
248 2019-12-26  91.599998  91.720001  91.070000  91.339996  89.228210   
249 2019-12-27  91.680000  91.699997  91.269997  91.500000  89.384514   

       Volume  
0    14427600  
1    13273500  
2    14077400  
3    15071700  
4    11351800  
..        ...  
245  17032100  
246   7534300  
247   2235900  
248   2922400  
249   3586800  

[250 rows x 7 columns]

MSFT


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   99.550003  101.750000   98.940002  101.120003   98.602066   
1   2019-01-03  100.099998  100.190002   97.199997   97.400002   94.974693   
2   2019-01-04   99.720001  102.510002   98.930000  101.930000   99.391899   
3   2019-01-07  101.639999  103.269997  100.980003  102.059998   99.518669   
4   2019-01-08  103.040001  103.970001  101.709999  102.800003  100.240234   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  157.350006  158.490005  156.289993  157.410004  155.759949   
246 2019-12-23  158.119995  158.119995  157.270004  157.410004  155.759949   
247 2019-12-24  157.479996  157.710007  157.119995  157.380005  155.730255   
248 2019-12-26  157.559998  158.729996  157.399994  158.669998  157.006729   
249 2019-12-27  159.449997  159.550003  158.220001  158.960007  157.293686   

       Volume  
0    35329300  
1    42579100  
2    44060600  
3    35656100  
4    31514400  
..        ...  
245  53477500  
246  17718200  
247   8989200  
248  14520600  
249  18412800  

[250 rows x 7 columns]

NKE


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   72.790001   74.639999   72.190002   74.059998   72.738121   
1   2019-01-03   73.250000   73.320000   71.209999   72.750000   71.451500   
2   2019-01-04   73.449997   75.120003   73.120003   74.650002   73.317596   
3   2019-01-07   74.739998   76.360001   74.300003   75.720001   74.368500   
4   2019-01-08   76.830002   77.400002   76.199997   76.730003   75.360481   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  101.000000  101.400002   98.849998   99.959999   99.215599   
246 2019-12-23  100.230003  100.690002   99.550003  100.040001   99.295006   
247 2019-12-24   99.830002  100.660004   99.760002  100.139999   99.394257   
248 2019-12-26  100.339996  100.750000  100.110001  100.709999   99.960014   
249 2019-12-27  101.000000  101.790001  100.989998  101.570000  100.813614   

       Volume  
0     6762700  
1     8007400  
2     7844200  
3     8184800  
4     8809000  
..        ...  
245  17641600  
246   6350500  
247   3328100  
248   3709900  
249   5307500  

[250 rows x 7 columns]

PFE


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  40.910816  41.271347  40.455406  41.034157  38.012341   
1   2019-01-03  41.024670  41.119545  39.800758  39.886147  36.948875   
2   2019-01-04  40.265656  41.129032  40.094875  40.796963  37.792622   
3   2019-01-07  40.863377  41.385201  40.664135  41.015179  37.994766   
4   2019-01-08  41.328274  41.461102  40.939278  41.204933  38.170547   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  37.333965  37.514233  36.973434  37.220116  35.759350   
246 2019-12-23  37.400379  37.495255  37.125237  37.305504  35.841389   
247 2019-12-24  37.314991  37.419353  37.191650  37.277039  35.814041   
248 2019-12-26  37.286526  37.542694  37.191650  37.362431  35.896080   
249 2019-12-27  37.409866  37.590134  37.239090  37.305504  35.841389   

       Volume  
0    26430300  
1    28503500  
2    27145300  
3    20995500  
4    19677200  
..        ...  
245  58199800  
246  17454300  
247   5187700  
248   9384100  
249  10117700  

[250 rows x 7 columns]

PG


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   91.029999   91.389999   89.930000   91.279999   86.643158   
1   2019-01-03   90.940002   92.500000   90.379997   90.639999   86.035660   
2   2019-01-04   90.839996   92.489998   90.370003   92.489998   87.791695   
3   2019-01-07   91.910004   92.500000   91.400002   92.120003   87.440491   
4   2019-01-08   92.699997   93.089996   91.820000   92.459999   87.763199   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  126.150002  126.599998  125.029999  125.360001  122.338455   
246 2019-12-23  125.519997  125.820000  124.690002  124.900002  121.889549   
247 2019-12-24  124.790001  125.309998  124.580002  125.220001  122.201820   
248 2019-12-26  125.099998  125.750000  124.870003  125.220001  122.201820   
249 2019-12-27  125.559998  126.300003  125.389999  126.089996  123.050858   

       Volume  
0     9843900  
1     9820200  
2    10565700  
3     9012500  
4     9551300  
..        ...  
245  16250500  
246   6619500  
247   2022100  
248   4760400  
249   5192000  

[250 rows x 7 columns]

TRV


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  117.489998  118.190002  116.330002  117.370003  112.364258   
1   2019-01-03  117.050003  117.300003  115.059998  115.260002  110.344231   
2   2019-01-04  116.660004  118.779999  116.129997  118.680000  113.618393   
3   2019-01-07  118.209999  119.540001  117.709999  118.300003  113.254585   
4   2019-01-08  119.379997  119.510002  116.839996  117.839996  112.814194   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  137.149994  137.610001  136.039993  136.429993  133.627701   
246 2019-12-23  136.800003  136.800003  135.580002  136.229996  133.431793   
247 2019-12-24  136.449997  136.449997  135.550003  135.830002  133.040024   
248 2019-12-26  135.960007  136.309998  135.809998  136.289993  133.490570   
249 2019-12-27  136.479996  136.899994  135.970001  136.800003  133.990112   

      Volume  
0    1793400  
1    1887500  
2    1793900  
3    1082800  
4    1403900  
..       ...  
245  2676500  
246  1003000  
247   486200  
248   927700  
249   614400  

[250 rows x 7 columns]

UNH


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  245.000000  246.350006  241.470001  243.490005  236.411499   
1   2019-01-03  243.479996  243.479996  236.130005  236.850006  229.964493   
2   2019-01-04  239.800003  243.460007  238.800003  239.619995  232.653992   
3   2019-01-07  240.399994  241.979996  238.229996  240.080002  233.100616   
4   2019-01-08  242.330002  244.789993  239.710007  243.289993  236.217270   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  298.899994  300.000000  292.510010  292.589996  288.888336   
246 2019-12-23  293.429993  295.850006  292.809998  295.089996  291.356720   
247 2019-12-24  295.260010  295.829987  293.559998  294.540009  290.813660   
248 2019-12-26  295.299988  296.260010  294.540009  295.649994  291.909607   
249 2019-12-27  296.010010  296.540009  295.010010  295.970001  292.225555   

      Volume  
0    4063600  
1    4623200  
2    5367600  
3    4133000  
4    3618600  
..       ...  
245  6608300  
246  2098900  
247   714000  
248  1050600  
249  1547100  

[250 rows x 7 columns]

RTX


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  66.179985  68.187538  65.601006  68.156075  64.796608   
1   2019-01-03  67.684074  67.967278  64.965385  65.122719  61.912777   
2   2019-01-04  65.525490  67.463814  65.141602  67.350533  64.030777   
3   2019-01-07  67.375710  67.595970  66.419128  67.319069  64.000862   
4   2019-01-08  68.495911  68.879799  67.558212  68.533669  65.155594   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  94.965385  94.965385  93.895531  94.078033  91.460083   
246 2019-12-23  94.373817  94.745125  94.122086  94.285713  91.661995   
247 2019-12-24  94.443047  94.600380  94.046570  94.128380  91.509033   
248 2019-12-26  94.367523  94.877281  94.241661  94.845818  92.206505   
249 2019-12-27  95.028320  95.072372  94.430458  94.575203  91.943428   

       Volume  
0     9361700  
1     8884200  
2    10083300  
3     8314600  
4     5951200  
..        ...  
245   7252600  
246   4361300  
247   1596700  
248   2231200  
249   2928000  

[250 rows x 7 columns]

VZ


Date       Open       High        Low      Close  Adj Close  \
0   2019-01-02  56.160000  56.160000  55.279999  56.020000  51.492947   
1   2019-01-03  56.299999  56.990002  56.060001  56.220001  51.676788   
2   2019-01-04  56.490002  56.889999  55.950001  56.360001  51.805481   
3   2019-01-07  56.759998  57.259998  56.549999  56.720001  52.136379   
4   2019-01-08  57.270000  58.689999  57.049999  58.380001  53.662239   
..         ...        ...        ...        ...        ...        ...   
245 2019-12-20  61.459999  62.220001  61.139999  62.070000  59.468941   
246 2019-12-23  62.009998  62.150002  61.320000  61.400002  58.827019   
247 2019-12-24  61.459999  61.459999  61.020000  61.279999  58.712044   
248 2019-12-26  61.279999  61.369999  61.110001  61.290001  58.721626   
249 2019-12-27  61.389999  61.619999  61.250000  61.529999  58.951569   

       Volume  
0    13969700  
1    19245400  
2    17035800  
3    16349000  
4    21424600  
..        ...  
245  42977900  
246   9477500  
247   6640100  
248   8117800  
249   8166800  

[250 rows x 7 columns]

V


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  130.000000  133.740005  129.600006  132.919998  131.264542   
1   2019-01-03  131.210007  131.279999  127.879997  128.130005  126.534225   
2   2019-01-04  130.440002  134.589996  130.130005  133.649994  131.985474   
3   2019-01-07  134.449997  137.169998  134.039993  136.059998  134.365448   
4   2019-01-08  137.970001  138.190002  135.960007  136.800003  135.096252   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  187.600006  188.169998  186.300003  188.000000  186.849869   
246 2019-12-23  188.149994  188.410004  187.000000  187.080002  185.935501   
247 2019-12-24  187.100006  188.020004  186.910004  187.570007  186.422501   
248 2019-12-26  187.889999  189.179993  187.699997  189.160004  188.002792   
249 2019-12-27  189.740005  189.889999  188.470001  189.389999  188.231369   

       Volume  
0     8788000  
1     9428300  
2    11065800  
3    12928000  
4     9243000  
..        ...  
245  11259500  
246   5831700  
247   2420900  
248   5237000  
249   5448600  

[250 rows x 7 columns]

WBA


Date       Open       High        Low      Close  Adj Close   Volume
0   2019-01-02  67.199997  68.459999  66.690002  68.040001  63.170071  5528100
1   2019-01-03  67.800003  68.089996  66.720001  67.330002  62.510887  5853200
2   2019-01-04  68.430000  70.139999  68.000000  69.570000  64.590569  6868600
3   2019-01-07  69.150002  70.540001  68.820000  69.970001  64.961929  6192500
4   2019-01-08  70.519997  71.669998  70.139999  71.370003  66.261726  6548100
..         ...        ...        ...        ...        ...        ...      ...
245 2019-12-20  58.330002  58.980000  57.759998  58.389999  55.920170  8433100
246 2019-12-23  58.709999  59.000000  58.419998  58.570000  56.092560  4813200
247 2019-12-24  58.900002  58.919998  58.279999  58.349998  55.881863  1475400
248 2019-12-26  58.340000  59.020000  58.299999  58.900002  56.408600  2736100
249 2019-12-27  58.919998  59.029999  58.380001  59.020000  56.523521  3647600

[250 rows x 7 columns]

WMT


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02   91.639999   93.650002   91.639999   93.339996   90.313919   
1   2019-01-03   93.209999   94.709999   92.699997   92.860001   89.849495   
2   2019-01-04   93.209999   93.660004   92.690002   93.440002   90.410690   
3   2019-01-07   93.620003   95.169998   93.180000   94.540001   91.475037   
4   2019-01-08   95.059998   95.760002   93.910004   95.199997   92.113625   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  121.480003  122.120003  120.019997  120.290001  118.747406   
246 2019-12-23  120.199997  120.690002  119.010002  119.029999  117.503563   
247 2019-12-24  119.129997  119.599998  119.010002  119.510002  117.977409   
248 2019-12-26  119.610001  119.849998  118.949997  119.519997  117.987274   
249 2019-12-27  119.940002  119.940002  119.250000  119.589996  118.056381   

      Volume  
0    8152700  
1    8277300  
2    8029100  
3    7789700  
4    7200900  
..       ...  
245  7954400  
246  4485800  
247  2227400  
248  4223800  
249  3544000  

[250 rows x 7 columns]

DIS


Date        Open        High         Low       Close   Adj Close  \
0   2019-01-02  108.099998  109.139999  107.730003  108.970001  107.654343   
1   2019-01-03  108.480003  108.650002  105.940002  106.330002  105.046211   
2   2019-01-04  107.940002  110.750000  107.250000  109.610001  108.286606   
3   2019-01-07  109.910004  111.400002  109.300003  110.559998  109.225136   
4   2019-01-08  111.800003  112.559998  111.169998  111.419998  110.074753   
..         ...         ...         ...         ...         ...         ...   
245 2019-12-20  147.149994  147.880005  145.770004  146.880005  146.880005   
246 2019-12-23  145.910004  146.330002  144.330002  144.679993  144.679993   
247 2019-12-24  144.580002  145.429993  144.449997  145.289993  145.289993   
248 2019-12-26  145.399994  145.860001  145.169998  145.699997  145.699997   
249 2019-12-27  146.050003  146.509995  145.449997  145.750000  145.750000   

       Volume  
0     9723500  
1    10594700  
2    10122800  
3     6714700  
4     8730700  
..        ...  
245  12629900  
246   9314000  
247   3508500  
248   4422000  
249   5495300  

[250 rows x 7 columns]

DOW


Date       Open       High        Low      Close  Adj Close  \
0   2019-03-20  52.750000  53.500000  49.500000  49.799999  45.368008   
1   2019-03-21  49.990002  50.000000  48.200001  48.980000  44.620983   
2   2019-03-22  48.799999  49.950001  48.160000  48.599998  44.274803   
3   2019-03-25  48.599998  49.400002  48.000000  49.150002  44.775864   
4   2019-03-26  49.000000  49.750000  48.180000  48.849998  44.502556   
..         ...        ...        ...        ...        ...        ...   
192 2019-12-20  55.049999  55.250000  54.020000  54.610001  51.977295   
193 2019-12-23  54.490002  55.049999  54.220001  54.930000  52.281868   
194 2019-12-24  55.000000  55.250000  54.650002  54.900002  52.253315   
195 2019-12-26  54.980000  55.230000  54.520000  55.220001  52.557884   
196 2019-12-27  55.290001  55.549999  54.720001  54.830002  52.186687   

       Volume  
0     2350800  
1     1764700  
2      844700  
3      440900  
4      504700  
..        ...  
192  10476700  
193   2678600  
194    926200  
195   1909600  
196   1688500  

[197 rows x 7 columns]

In [109]:
data['DOW']['table'].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198 entries, 0 to 197
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       198 non-null    datetime64[ns]
 1   Open       198 non-null    float64       
 2   High       198 non-null    float64       
 3   Low        198 non-null    float64       
 4   Close      198 non-null    float64       
 5   Adj Close  198 non-null    float64       
 6   Volume     198 non-null    int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 11.0 KB


<br><br>

### Questions

<br><br>

**Question 1.** What is the average change of price over the year (in %)?

*Note 1*. The opening price is the price at which a stock first trades upon the opening of an exchange on a trading day.

*Note 2*. The closing price for any stock is the final price at which it trades during regular market hours on any given day.

*Note 3*. Here by the price change we going to mean a ratio of a closing price in the last day of the period to an opening price in the first day of that period, subtracted one and multiplied by 100.

Example: if a price of a stock in day 1 opened at \\$100, and its close price in the last day was \\$120, then the price change during the period is: $$\left(\dfrac{120}{100}-1\right) * 100 = (1.2 - 1) * 100=20.$$

The price grew by 20%.

In [166]:
avg = 0
avg0 = 0
for ticker in dow_jones_companies:
    df = data[ticker]['table']
    df['Change'] = df.apply(lambda row: 100*(row.Close/row.Open - 1), axis=1) # Create new column with 100*(Close/Open - 1)
    data[ticker]['VolumeProgress'] = df.Change.sum()
    avg0 += data[ticker]['VolumeProgress']

    mindate = min(df.Date) # Min date
    maxdate = max(df.Date) # Max date
    opn = df[df.Date == mindate].Open.values[0]
    cls = df[df.Date == maxdate].Close.values[0]
    data[ticker]['Open' ] = opn
    data[ticker]['Close'] = cls
#     display(df[df.Date == mindate])
#     display(df[df.Date == maxdate])
    data[ticker]['Progress'] = (cls/opn - 1)*100

    print('Ticker:',ticker,'Date:', mindate,'Open:',opn,'Date:', maxdate,', Close:',cls, 'Progress:', data[ticker]['Progress'])    
    avg += data[ticker]['Progress']*(maxdate-mindate)/( pd.to_datetime('2019-12-30') - pd.to_datetime('2019-01-02'))
    
answer_part_1 = avg/N
print('AVG:',answer_part_1)

# I've calculated % change between close at 31st of December and open in 1st day of Jan, this worked

Ticker: AXP Date: 2019-01-02 00:00:00 Open: 93.910004 Date: 2019-12-30 00:00:00 , Close: 124.300003 Progress: 32.36076850768743
Ticker: AAPL Date: 2019-01-02 00:00:00 Open: 38.7225 Date: 2019-12-30 00:00:00 , Close: 72.879997 Progress: 88.21098069597781
Ticker: BA Date: 2019-01-02 00:00:00 Open: 316.190002 Date: 2019-12-30 00:00:00 , Close: 326.399994 Progress: 3.2290685775700156
Ticker: CAT Date: 2019-01-02 00:00:00 Open: 124.029999 Date: 2019-12-30 00:00:00 , Close: 147.520004 Progress: 18.93897056308127
Ticker: CSCO Date: 2019-01-02 00:00:00 Open: 42.279999 Date: 2019-12-30 00:00:00 , Close: 47.59 Progress: 12.55913227434089
Ticker: CVX Date: 2019-01-02 00:00:00 Open: 107.339996 Date: 2019-12-30 00:00:00 , Close: 119.849998 Progress: 11.654557915206176
Ticker: XOM Date: 2019-01-02 00:00:00 Open: 67.349998 Date: 2019-12-30 00:00:00 , Close: 69.480003 Progress: 3.162591036751028
Ticker: GS Date: 2019-01-02 00:00:00 Open: 164.330002 Date: 2019-12-30 00:00:00 , Close: 229.80000299999998

In [ ]:
# Ticker:
#     Date: 2019-01-02 00:00:00 Open: 93.910004 
#     Date: 2019-12-30 00:00:00 Close: 124.300003 
#     Progress: 32.36076850768743

<br>

**Question 2.** What company's stock price grew the most (in %)? Enter ticker of the company as an answer)

In [123]:
maxGrew = -100000
maxGrewCompany = ''
for ticker in dow_jones_companies:
    progress = data[ticker]['Progress']
    print('Ticker:', ticker, 'progress:', progress)
    if maxGrew < progress:
        maxGrew = progress
        maxGrewCompany = ticker
answer_part_2 = maxGrewCompany
print(answer_part_2)

Ticker: AXP progress: 32.36076850768743
Ticker: AAPL progress: 88.21098069597781
Ticker: BA progress: 3.2290685775700156
Ticker: CAT progress: 18.93897056308127
Ticker: CSCO progress: 12.55913227434089
Ticker: CVX progress: 11.654557915206176
Ticker: XOM progress: 3.162591036751028
Ticker: GS progress: 39.840564840983795
Ticker: HD progress: 28.047839866036874
Ticker: IBM progress: 18.569766653517263
Ticker: INTC progress: 29.72149760055478
Ticker: JNJ progress: 13.40045058142314
Ticker: KO progress: 17.746061306903748
Ticker: JPM progress: 44.48151051010456
Ticker: MCD progress: 12.256997611151066
Ticker: MMM progress: -6.383774120506769
Ticker: MRK progress: 20.905827853555216
Ticker: MSFT progress: 58.30235183418324
Ticker: NKE progress: 38.480562735532864
Ticker: PFE progress: -9.763452286065366
Ticker: PG progress: 36.735144861420885
Ticker: TRV progress: 16.17159530464882
Ticker: UNH progress: 19.938777959183685
Ticker: RTX progress: 42.52567600309971
Ticker: VZ progress: 8.99216

<br>

**Question 3.** What company's stock lost in price the most (in %)? Enter ticker of the company as an answer

In [125]:
minGrew = 100000
minGrewCompany = ''
for ticker in dow_jones_companies:
    progress = data[ticker]['Progress']
    print('Ticker:', ticker, 'progress:', progress)
    if minGrew > progress:
        minGrew = progress
        minGrewCompany = ticker
answer_part_3 = minGrewCompany
print(answer_part_3)

Ticker: AXP progress: 32.36076850768743
Ticker: AAPL progress: 88.21098069597781
Ticker: BA progress: 3.2290685775700156
Ticker: CAT progress: 18.93897056308127
Ticker: CSCO progress: 12.55913227434089
Ticker: CVX progress: 11.654557915206176
Ticker: XOM progress: 3.162591036751028
Ticker: GS progress: 39.840564840983795
Ticker: HD progress: 28.047839866036874
Ticker: IBM progress: 18.569766653517263
Ticker: INTC progress: 29.72149760055478
Ticker: JNJ progress: 13.40045058142314
Ticker: KO progress: 17.746061306903748
Ticker: JPM progress: 44.48151051010456
Ticker: MCD progress: 12.256997611151066
Ticker: MMM progress: -6.383774120506769
Ticker: MRK progress: 20.905827853555216
Ticker: MSFT progress: 58.30235183418324
Ticker: NKE progress: 38.480562735532864
Ticker: PFE progress: -9.763452286065366
Ticker: PG progress: 36.735144861420885
Ticker: TRV progress: 16.17159530464882
Ticker: UNH progress: 19.938777959183685
Ticker: RTX progress: 42.52567600309971
Ticker: VZ progress: 8.99216

<br>

**Question 4.** What company had the largest summary volume over the year? Enter ticker of the company as an answer

In [143]:
maxVolume = -100000
maxVolumeCompany = ''
for ticker in dow_jones_companies:
    df = data[ticker]['table']
    data[ticker]['SummaryVolume'] = df.Volume.sum()
    if data[ticker]['SummaryVolume'] >maxVolume:
        maxVolume = data[ticker]['SummaryVolume']
        maxVolumeCompany = ticker
answer_part_4 = maxVolumeCompany
print('Max Volume company:',answer_part_4)

Max Volume company: AAPL


<br>

**Question 5.** What is the biggest stock price daily increase (in %)? Enter the number 

In [177]:
maxVol = np.zeros(N)
maxTick = ''
for i,ticker in enumerate(dow_jones_companies):
    df = data[ticker]['table']
    df['Change'] = df.apply(lambda row: 100*(row.Close/row.Open - 1), axis=1) # Create new column with 100*(Close/Open - 1)
    
    maxVol[i] = df['Change'].max()
arg = np.argmax(maxVol)
    
print(maxVol)        
answer_part_5 = maxVol.max()

[2.38019915 3.9548283  7.741003   3.6736817  3.05878017 3.70212138
 3.47439357 5.3903754  2.40970404 5.00351228 4.18896483 2.8114301
 2.30394555 3.50182502 1.86248959 2.5856949  2.3598739  2.78992409
 2.92559888 3.21518838 2.16033146 2.82369314 5.40386646 3.64369296
 3.10899963 3.51576559 5.58404157 2.15517438 7.31707234 4.38810831]


<br><br>

**Question 6.** What is the company that had the biggest stock price daily increase? Enter ticker of the company as an answer

In [180]:
print(dow_jones_companies[arg]) 
answer_part_6 = dow_jones_companies[arg]

BA


<br>

**Question 7.** What is the biggest stock price daily decrease (in %)? Enter the number

In [196]:
minVol = np.zeros(N)

for i,ticker in enumerate(dow_jones_companies):
    df = data[ticker]['table']
    df['Change'] = df.apply(lambda row: 100*(row.Close/row.Open - 1), axis=1) # Create new column with 100*(Close/Open - 1)
    
    minVol[i] = df['Change'].min()
arg_min = np.argmin(minVol)
    
print(minVol)        
answer_part_7 = minVol.min()


[-3.4754082  -3.24212971 -6.77506775 -3.94819382 -3.84692834 -2.73329008
 -2.65424954 -3.53316608 -2.25119539 -3.03823831 -3.59696626 -4.20823725
 -2.37687575 -2.40709371 -2.83777469 -4.74840396 -4.75454574 -3.23778931
 -3.53285811 -3.07911538 -3.91381381 -2.74689154 -7.16355966 -5.52767242
 -3.23936783 -3.46892599 -5.98447018 -3.1701413  -2.8337371  -5.59241896]


<br>

**Question 8.** What is the company that had the biggest stock price daily decrese (in %)? Enter ticker of the company as an answer

In [199]:
answer_part_8 = dow_jones_companies[arg_min]